File: sdk/npm/src/providers/mffac.ts
import { InternalEmailInfo, Email, Channel } from '../types';
import { normalizeEmail } from '../normalize';
import { fetchWithTimeout } from '../retry';

const CHANNEL: Channel = 'mffac';
const BASE_URL = 'https://www.mffac.com/api';

const DEFAULT_HEADERS = {
  'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/146.0.0.0 Safari/537.36 Edg/146.0.0.0',
  'Content-Type': 'application/json',
  'Accept': '*/*',
  'Origin': 'https://www.mffac.com',
  'Referer': 'https://www.mffac.com/',
  'sec-ch-ua': '"Chromium";v="146", "Not-A.Brand";v="24", "Microsoft Edge";v="146"',
  'sec-ch-ua-mobile': '?0',
  'sec-ch-ua-platform': '"Windows"',
  'sec-fetch-dest': 'empty',
  'sec-fetch-mode': 'cors',
  'sec-fetch-site': 'same-origin',
};

const GET_HEADERS = Object.fromEntries(
  Object.entries(DEFAULT_HEADERS).filter(([key]) => key.toLowerCase() !== 'content-type'),
) as Record<string, string>;

function receivedAtToIso(value: unknown): string {
  const seconds = typeof value === 'number' ? value : Number(value);
  if (!Number.isFinite(seconds) || seconds <= 0) return '';
  return new Date(seconds * 1000).toISOString();
}

function flattenEmail(raw: any, recipient: string): Record<string, unknown> {
  return {
    id: raw.id,
    from: raw.fromAddress,
    to: raw.toAddress || recipient,
    subject: raw.subject || '',
    text: raw.textContent || '',
    html: raw.htmlContent || '',
    date: receivedAtToIso(raw.receivedAt),
    isRead: raw.isRead,
    attachments: [],
  };
}

async function fetchEmailDetail(id: string): Promise<any | null> {
  const response = await fetchWithTimeout(`${BASE_URL}/emails/${encodeURIComponent(id)}`, {
    method: 'GET',
    headers: GET_HEADERS,
  });
  if (!response.ok) return null;
  const data = await response.json();
  return data && data.success && data.email && typeof data.email === 'object' ? data.email : null;
}

export async function generateEmail(): Promise<InternalEmailInfo> {
  const response = await fetchWithTimeout(`${BASE_URL}/mailboxes`, {
    method: 'POST',
    headers: DEFAULT_HEADERS,
    body: JSON.stringify({ expiresInHours: 24 }),
  });

  if (!response.ok) {
    throw new Error(`Failed to generate email: ${response.status}`);
  }

  const data = await response.json();

  if (!data.success || !data.mailbox) {
    throw new Error('Failed to generate email');
  }

  const mailbox = data.mailbox;
  const email = `${mailbox.address}@mffac.com`;

  return {
    channel: CHANNEL,
    email,
    expiresAt: mailbox.expiresAt,
    createdAt: mailbox.createdAt,
    token: mailbox.id,
  };
}

export async function getEmails(email: string, _token?: string): Promise<Email[]> {
  const address = email.split('@')[0];
  const response = await fetchWithTimeout(`${BASE_URL}/mailboxes/${address}/emails`, {
    method: 'GET',
    headers: GET_HEADERS,
  });

  if (!response.ok) {
    throw new Error(`Failed to get emails: ${response.status}`);
  }

  const data = await response.json();

  if (!data.success) {
    throw new Error('Failed to get emails');
  }

  const rawEmails = data.emails || [];
  const emails: Email[] = [];
  for (const raw of rawEmails) {
    const id = String(raw?.id || '').trim();
    const detail = id ? await fetchEmailDetail(id) : null;
    emails.push(normalizeEmail(flattenEmail(detail || raw, email), email));
  }
  return emails;
}

File: sdk/go/provider/mffac.go
package provider

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"strings"

	http "github.com/bogdanfinn/fhttp"
)

const mffacAPIBase = "https://www.mffac.com/api"

func mffacDefaultHeaders(req *http.Request) {
	req.Header.Set("User-Agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/146.0.0.0 Safari/537.36 Edg/146.0.0.0")
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Accept", "*/*")
	req.Header.Set("Origin", "https://www.mffac.com")
	req.Header.Set("Referer", "https://www.mffac.com/")
	req.Header.Set("sec-ch-ua", `"Chromium";v="146", "Not-A.Brand";v="24", "Microsoft Edge";v="146"`)
	req.Header.Set("sec-ch-ua-mobile", "?0")
	req.Header.Set("sec-ch-ua-platform", `"Windows"`)
	req.Header.Set("sec-fetch-dest", "empty")
	req.Header.Set("sec-fetch-mode", "cors")
	req.Header.Set("sec-fetch-site", "same-origin")
}

func MffacGenerate() (*CreatedMailbox, error) {
	req, err := http.NewRequest("POST", mffacAPIBase+"/mailboxes", bytes.NewReader([]byte(`{"expiresInHours":24}`)))
	if err != nil {
		return nil, err
	}
	mffacDefaultHeaders(req)

	resp, err := HTTPClient().Do(req)
	if err != nil {
		return nil, fmt.Errorf("mffac generate: %w", err)
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return nil, fmt.Errorf("mffac generate: %d", resp.StatusCode)
	}

	var parsed struct {
		Success bool `json:"success"`
		Mailbox *struct {
			Address   string `json:"address"`
			ID        string `json:"id"`
			ExpiresAt any    `json:"expiresAt"`
			CreatedAt string `json:"createdAt"`
		} `json:"mailbox"`
	}
	if err := json.Unmarshal(body, &parsed); err != nil {
		return nil, err
	}
	if !parsed.Success || parsed.Mailbox == nil || parsed.Mailbox.Address == "" || parsed.Mailbox.ID == "" {
		return nil, fmt.Errorf("mffac: invalid response")
	}

	email := parsed.Mailbox.Address + "@mffac.com"
	info := &CreatedMailbox{Channel: "mffac", Email: email, Token: parsed.Mailbox.ID}
	info.ExpiresAt = parsed.Mailbox.ExpiresAt
	info.CreatedAt = parsed.Mailbox.CreatedAt
	return info, nil
}

func MffacGetEmails(email string, _token string) ([]NormEmail, error) {
	local := email
	if i := strings.LastIndex(email, "@"); i > 0 {
		local = email[:i]
	}
	u := fmt.Sprintf("%s/mailboxes/%s/emails", mffacAPIBase, local)
	req, err := http.NewRequest("GET", u, nil)
	if err != nil {
		return nil, err
	}
	mffacDefaultHeaders(req)
	req.Header.Del("Content-Type")

	resp, err := HTTPClient().Do(req)
	if err != nil {
		return nil, fmt.Errorf("mffac emails: %w", err)
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return nil, fmt.Errorf("mffac emails: %d", resp.StatusCode)
	}

	var parsed struct {
		Success bool            `json:"success"`
		Emails  json.RawMessage `json:"emails"`
	}
	if err := json.Unmarshal(body, &parsed); err != nil {
		return nil, err
	}
	if !parsed.Success {
		return nil, fmt.Errorf("mffac: success false")
	}

	var rawList []json.RawMessage
	if len(parsed.Emails) > 0 && string(parsed.Emails) != "null" {
		if err := json.Unmarshal(parsed.Emails, &rawList); err != nil {
			return nil, err
		}
	}
	return NormalizeRawMessages(rawList, email)
}
